File: app/src/pfp/java/social/entourage/android/map/entourage/create/CreateEntourageFragment.java
package social.entourage.android.map.entourage.create;

import android.graphics.Typeface;
import android.os.Bundle;
import android.support.v4.content.res.ResourcesCompat;
import android.view.View;
import android.widget.Switch;
import android.widget.TextView;

import com.google.android.gms.location.places.Place;
import com.google.android.gms.maps.model.LatLng;

import java.text.DateFormat;
import java.text.SimpleDateFormat;
import java.util.Locale;

import butterknife.BindView;
import butterknife.OnClick;
import social.entourage.android.Constants;
import social.entourage.android.EntourageComponent;
import social.entourage.android.EntourageEvents;
import social.entourage.android.R;
import social.entourage.android.api.model.map.BaseEntourage;
import social.entourage.android.api.model.map.Entourage;
import social.entourage.android.location.LocationFragment;

/**
 *
 */
public class CreateEntourageFragment extends BaseCreateEntourageFragment implements LocationFragment.OnFragmentInteractionListener, CreateEntourageListener {

    // ----------------------------------
    // Constants
    // ----------------------------------

    // ----------------------------------
    // Attributes
    // ----------------------------------

    @BindView(R.id.create_entourage_title_label)
    TextView titleLabel;

    @BindView(R.id.create_entourage_description_label)
    TextView descriptionLabel;

    @BindView(R.id.create_entourage_position_description)
    TextView positionLabel;

    @BindView(R.id.create_entourage_date_label)
    TextView dateLabel;

    // ----------------------------------
    // Lifecycle
    // ----------------------------------

    public CreateEntourageFragment() {
        // Required empty public constructor
    }

    protected void setupComponent(EntourageComponent entourageComponent) {
        DaggerCreateEntourageComponent.builder()
                .entourageComponent(entourageComponent)
                .createEntourageModule(new CreateEntourageModule(this))
                .build()
                .inject(this);
    }

    // ----------------------------------
    // Interactions handling
    // ----------------------------------

    // ----------------------------------
    // Presenter callbacks
    // ----------------------------------

    // ----------------------------------
    // Base class overrides
    // ----------------------------------

    @Override
    protected void initializeLocation() {
        Bundle args = getArguments();
        if (args != null) {
            if (editedEntourage != null) {
                location = editedEntourage.getLocation().getLocation();
                BaseEntourage.Metadata metadata = editedEntourage.getMetadata();
                if (metadata != null) {
                    positionTextView.setText(metadata.getDisplayAddress());
                    positionLabel.setVisibility(View.GONE);
                }
            } else {
                location = args.getParcelable(KEY_ENTOURAGE_LOCATION);
            }
        }
    }

    @Override
    protected void initializeTitleEditText() {
        if (editedEntourage != null) {
            onTitleChanged(editedEntourage.getTitle());
        }
    }

    @Override
    protected void initializeDescriptionEditText() {
        if (editedEntourage != null) {
            onDescriptionChanged(editedEntourage.getDescription());
        }
    }

    @Override
    protected void initializeCategory() {
        if (groupType == null) groupType = Entourage.TYPE_OUTING; // only outings
        super.initializeCategory();
    }

    @Override
    protected void updateDateTextView() {
        super.updateDateTextView();
        dateLabel.setVisibility(View.GONE);
    }

    @Override
    protected boolean isValid() {
        joinRequestTypePublic = privacySwitch.isChecked();
        return super.isValid();
    }

    // ----------------------------------
    // LocationFragment.OnFragmentInteractionListener
    // ----------------------------------

    @Override
    public void onEntourageLocationChosen(LatLng location, String address, Place place) {
        // for PFP, we use only the place
        if (place != null) {
            this.location = place.getLatLng();
            positionTextView.setText(place.getAddress());
            positionLabel.setVisibility(View.GONE);
            if (groupType != null && groupType.equalsIgnoreCase(Entourage.TYPE_OUTING)) {
                if (entourageMetadata == null) entourageMetadata = new BaseEntourage.Metadata();
                entourageMetadata.setPlaceName(place.getName().toString());
                if (place.getAddress() != null) {
                    entourageMetadata.setStreetAddress(place.getAddress().toString());
                }
                entourageMetadata.setGooglePlaceId(place.getId());
            }
        } else {
            this.location = null;
            positionTextView.setText("");
            positionLabel.setVisibility(View.VISIBLE);
        }
    }

    // ----------------------------------
    // CreateEntourageListener
    // ----------------------------------

    @Override
    public void onTitleChanged(final String title) {
        if (title == null || title.trim().length() == 0) {
            titleLabel.setVisibility(View.VISIBLE);
            titleEditText.setText("");
        } else {
            titleLabel.setVisibility(View.GONE);
            titleEditText.setText(title);
        }
    }

    @Override
    public void onDescriptionChanged(final String description) {
        if (description == null || description.trim().length() == 0) {
            descriptionLabel.setVisibility(View.VISIBLE);
            descriptionEditText.setText("");
        } else {
            descriptionLabel.setVisibility(View.GONE);
            descriptionEditText.setText(description);
        }
    }

}

File: app/src/pfp/java/social/entourage/android/entourage/create/CreateEntourageFragment.kt
package social.entourage.android.entourage.create

import android.view.View
import android.widget.TextView
import butterknife.BindView
import com.google.android.gms.maps.model.LatLng
import com.google.android.libraries.places.compat.Place
import social.entourage.android.EntourageComponent
import social.entourage.android.R
import social.entourage.android.api.model.map.BaseEntourage
import social.entourage.android.api.model.map.Entourage
import social.entourage.android.location.LocationFragment

/**
 *
 */
class CreateEntourageFragment  : BaseCreateEntourageFragment(), LocationFragment.OnFragmentInteractionListener, CreateEntourageListener {
    // ----------------------------------
    // Lifecycle
    // ----------------------------------
    // ----------------------------------
    // Constants
    // ----------------------------------
    // ----------------------------------
    // Attributes
    // ----------------------------------
    @JvmField
    @BindView(R.id.create_entourage_title_label)
    var titleLabel: TextView? = null

    @JvmField
    @BindView(R.id.create_entourage_description_label)
    var descriptionLabel: TextView? = null

    @JvmField
    @BindView(R.id.create_entourage_position_description)
    var positionLabel: TextView? = null

    @JvmField
    @BindView(R.id.create_entourage_date_start_label)
    var dateLabel: TextView? = null
    override fun setupComponent(entourageComponent: EntourageComponent) {
        DaggerCreateEntourageComponent.builder()
                .entourageComponent(entourageComponent)
                .createEntourageModule(CreateEntourageModule(this))
                .build()
                .inject(this)
    }

    // ----------------------------------
    // Interactions handling
    // ----------------------------------
    // ----------------------------------
    // Presenter callbacks
    // ----------------------------------
    // ----------------------------------
    // Base class overrides
    // ----------------------------------
    override fun initializeLocation() {
        val args = arguments
        if (args != null) {
            if (editedEntourage != null) {
                location = editedEntourage.location.location
                val metadata = editedEntourage.metadata
                if (metadata != null) {
                    positionTextView.text = metadata.displayAddress
                    positionLabel!!.visibility = View.GONE
                }
            } else {
                location = args.getParcelable(KEY_ENTOURAGE_LOCATION)
            }
        }
    }

    override fun initializeTitleEditText() {
        if (editedEntourage != null) {
            onTitleChanged(editedEntourage.title)
        }
    }

    override fun initializeDescriptionEditText() {
        if (editedEntourage != null) {
            onDescriptionChanged(editedEntourage.description)
        }
    }

    override fun initializeCategory() {
        if (groupType == null) groupType = BaseEntourage.TYPE_OUTING // only outings
        super.initializeCategory()
    }

    override fun updateDateStartTextView() {
        super.updateDateStartTextView()
        dateLabel!!.visibility = View.GONE
    }

    override fun updateDateEndTextView() {
        super.updateDateEndTextView()
        dateLabel!!.visibility = View.GONE
    }

    override fun isValid(): Boolean {
        joinRequestTypePublic = privacySwitch.isChecked
        return super.isValid()
    }

    // ----------------------------------
    // LocationFragment.OnFragmentInteractionListener
    // ----------------------------------
    override fun onEntourageLocationChosen(location: LatLng, address: String, place: Place) {
        // for PFP, we use only the place
        this.location = place.latLng
        positionTextView.text = place.address
        positionLabel!!.visibility = View.GONE
        if (groupType != null && groupType.equals(BaseEntourage.TYPE_OUTING, ignoreCase = true)) {
            if (entourageMetadata == null) entourageMetadata = BaseEntourage.Metadata()
            entourageMetadata.placeName = place.name.toString()
            if (place.address != null) {
                entourageMetadata.streetAddress = place.address.toString()
            }
            entourageMetadata.googlePlaceId = place.id
        }
    }

    // ----------------------------------
    // CreateEntourageListener
    // ----------------------------------
    override fun onTitleChanged(title: String) {
        if (title.trim { it <= ' ' }.isEmpty()) {
            titleLabel!!.visibility = View.VISIBLE
            titleEditText.text = ""
        } else {
            titleLabel!!.visibility = View.GONE
            titleEditText.text = title
        }
    }

    override fun onDescriptionChanged(description: String) {
        if (description.trim { it <= ' ' }.isEmpty()) {
            descriptionLabel!!.visibility = View.VISIBLE
            descriptionEditText.text = ""
        } else {
            descriptionLabel!!.visibility = View.GONE
            descriptionEditText.text = description
        }
    }
}